### ISSUE TITLE
CF check rejects many valid units

### ISSUE THREAD

**Issue by mhidas** [ioos/compliance-checker#81]
Running the CF tests on some [IMOS](http://imos.org.au/) files, we get high-priority tests failing with messages like these:

    units are metres, standard_name units should be m
    units are Celsius, standard_name units should be K
    
The CF convention says "The value of the units attribute is a string that can be recognized by UNIDATA's Udunits package" ([Section 3.1](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#units)) and "Unless it is dimensionless, a variable with a standard_name attribute must have units which are **physically equivalent (not necessarily identical) to the canonical units**" ([Section 3.3](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#standard-name)).

Looking at the code (cf.py), the function `check_units` seems to insist on the units being exactly the canonical units for any variable with a _standard_name_ attribute, but makes some exceptions to allow some commonly used variants. These exceptions include e.g. "meter" and "meters" (US spelling), but not "metres" (British/Australian spelling).

Wouldn't it be more correct (and more flexible) to check the units using the tests of equivalence provided by UDUNITS? (e.g. `udunitspy.udunits2.Unit` has a method `are_convertible`)

**Comment by dpsnowden** [comment]
:+1:

**Comment by ocefpaf** [comment]
Just an idea.  I know that compliance-checker crew wants to ditch udunits and use an online server (issue #62).  However, if we convince people from [iris](http://scitools.org.uk/iris/) to create a standalone Units package from the `iris.unit` module it will bring some advantages:

- `iris.unit` uses Udunits behind the scenes, so little change is needed to start using it here;
- `iris.unit` is available on windows, Mac, and Linux;
- sharing code between two projects leads to better maintenance of that software;
- No internet connect needed (for some people with flaky connections like myself that is a huge :+1: ).

Here is an example of what `iris.unit`can do:
http://nbviewer.ipython.org/gist/ocefpaf/5f6ae1fe1250312538df

**Comment by dpsnowden** [comment]
I agree that it's desirable to have a units module accessible locally.  I don't know the level of effort for this but it seems increasingly important.  udunitspy is a blocker for native windows installations of nearly all CF aware modules.  How much work would it take to 

1. Do it ourselves
2. Fund someone to do it
3. Beg iris developers to do it

Are there any collaborations we could participate in rather than lead?  How far along is the iris work?

**Comment by daf** [comment]
My initial thoughts for an interim solution are to make udunitspy an optional dep (not sure how that's best done with python packaging), then try/catch the import of it and use the web service as a secondary option with plenty of log messages to let the user know what's happening.  I like @ocefpaf's suggestion as a more polished solution.  Or simply a better udunitspy that does it natively instead of wrapping another library.

**Comment by ocefpaf** [comment]
@daf I am all for doing it natively, but udunits is a **de facto standard** and mimicking all of its functionally plus keeping up with its development will be a burden on the developers of such a library.  (Besides, python has enough [units libraries](https://conference.scipy.org/scipy2013/presentation_detail.php?id=174) :wink:)
  
@dpsnowden Lets ask iris devs and see what they have to say (I am asking right now in their mailing list).  If they are not interested or do not have the man power to do it you might consider adding the topic "fork `iris.units`" into options 1 or 2.

**Comment by ocefpaf** [comment]
> My initial thoughts for an interim solution are to make udunitspy an optional dep (not sure how that's best done with python packaging

Add this to `setup.py`

```python
 extras_require = dict(udunits="udunistpy")
```

Then install with:
```bash
pip install -e .[udunits] compliance-checker
```

PS: If using conda just add that to requirements.

**Comment by daf** [comment]
Isn't udunits just a C library on top of XML definitions, aka the XML defs are the canonical source of info?

**Comment by duncombe** [comment]
Minor nit-pick, pointing out the typos: your original post said "e.g.
"metre" and "metres" (US spelling), but not "metres" (British/Australian
spelling)" , should have read  "e.g. "meter" and "meters" (US spelling),
but not "metres" (British/Australian spelling)" !

On Wed, Feb 25, 2015 at 7:51 AM, Marty <notifications@github.com> wrote:

> Running the CF tests on some IMOS <http://imos.org.au/> files, we get
> high-priority tests failing with messages like these:
>
> units are metres, standard_name units should be m
> units are Celsius, standard_name units should be K
>
> The CF convention says "The value of the units attribute is a string that
> can be recognized by UNIDATA's Udunits package" (Section 3.1
> <http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#units>)
> and "Unless it is dimensionless, a variable with a standard_name attribute
> must have units which are *physically equivalent (not necessarily
> identical) to the canonical units*" (Section 3.3
> <http://cfconventions.org/Data/cf-conventions/cf-conventions-1.6/build/cf-conventions.html#standard-name>
> ).
>
> Looking at the code (cf.py), the function check_units seems to insist on
> the units being exactly the canonical units for any variable with a
> *standard_name* attribute, but makes some exceptions to allow some
> commonly used variants. These exceptions include e.g. "metre" and "metres"
> (US spelling), but not "metres" (British/Australian spelling).
>
> Wouldn't it be more correct (and more flexible) to check the units using
> the tests of equivalence provided by UDUNITS? (e.g.
> udunitspy.udunits2.Unit has a method are_convertible)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ioos/compliance-checker/issues/81>.
>



-- 


----------------------------------------------------------------
Dr Christopher M. Duncombe Rae                  Linux User #84980
@ Home: +27 21 531 9957                  @ Cell: +27 71 770
4848http://www.linkedin.com/in/christopherduncomberae
----------------------------------------------------------------

**Comment by mhidas** [comment]
@duncombe Yes, that's what I meant. Thanks!

**Comment by mhidas** [comment]
Thanks for looking at this issue. Most of the discussion above seems to actually refer to #62.
Could my issue be addressed independently of which units package is being used, i.e. using udunitspy for now?

**Comment by mhidas** [comment]
Just realised that this has actually been fixed in release 1.0.1 
We were using 1.0.0 when I reported this issue.
